testing/esp/esdlcmd/esdlcmd-test.py

```diff
@@ -0,0 +1,368 @@
+    """Settings for a specific test case."""
+
+    def __init__(self, run_settings, name, command, esdl_file, service, xsl_path, options=None):
+        self.run_settings = run_settings
+        self.name = name
+        self.command = command
+        self.esdl_path = (self.run_settings.test_path / 'inputs' / esdl_file)
+        self.service = service
+        self.xsl_path = xsl_path
+        self.options = options
+        self.output_path = Path(self.run_settings.output_base) / name
+        self.args = [
+            str(run_settings.exe_path),
+            self.command,
+            self.esdl_path,
+            self.service,
+            '--xslt',
+            self.xsl_path,
+            '--outdir',
+            # must contain a trailing slash
+            str(self.output_path) + '/'
+        ]
+
+        if options:
+            self.args.extend(options)
+
+        self.result = None
+
+
+    def run_test(self):
+        safe_mkdir(self.output_path)
+        logging.debug("Test %s args: %s", self.name, str(self.args))
+        self.result = subprocess.run(self.args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        if self.result.returncode != 0:
+            logging.error('Error running "esdl %s" for test "%s": %s', self.command, self.name, self.result.stderr)
+            success = False
+        else:
+            success = self.validate_results()
+
+        self.run_settings.stats.add_count(success)
+
+    def validate_results(self):
+        """Compare test case results to the known baseline.
+
+        Return True if the two are identical or False otherwise.
+        """
+        logging.debug('TestCaseBase implementation called, no comparison run')
+        return False
+
+
+class TestCaseXSD(TestCaseBase):
+    """Test case for the wsdl or xsd commands.
+
+    Both generate a single file output, so test validation compares
+    the contents of the output file with the baseline file.
+
+    The path the xsl files not include the 'xslt' directory. The command
+    assumes it needs to apped that directory itself.
+    """
+
+    def __init__(self, run_settings, name, command, esdl_file, service, xsl_path, options=None):
+        super().__init__(run_settings, name, command, esdl_file, service, xsl_path, options)
+
+    def validate_results(self):
+        """Compare test case results to the known baseline.
+
+        Return True if the two are identical or False otherwise.
+        """
+        suffix = '.' + self.command
+        outName = (self.output_path / self.service.lower()).with_suffix(suffix)
+        baseline = (self.run_settings.test_path / 'baselines' / self.name).with_suffix(suffix)
+
+        if (not baseline.exists()):
+            logging.error('Missing baseline file %s', str(baseline))
+            return False
+
+        if (not outName.exists()):
+            logging.error('Missing output for test %s', self.name)
+            return False
+
+        if (not filecmp.cmp(str(baseline), str(outName))):
```

Comment:
I don't know python, so can't comment too much. One question, though, is whether equality is necessary or equivalence might be sufficient? In other words, is there an expectation that one input will yield the exact same ordering and whitespace every time?

Reply:
I think that we should check for equality. Tony has mentioned that the wsdls and xsds generated should preserve order from the interface definitions.